Podcast Player: add colors setting to the block

retrofox opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

This PR adds the colors setting feature to the Podcast Player block.
It adds three kinds of colors that will be applied according to the following criteria.

1) Primary color (darker grey used in episode title at top, in the darker colored icons in the player, and in the selected episode in list below). 
2) Secondary color (light grey used in podcast name, most player controls, description, and non-selected episodes in the list below, and open in new tab)
3) Background color

<img width="931" alt="Screen Shot 2020-03-25 at 6 55 11 PM" src="https://user-images.githubusercontent.com/77539/77589600-bb1e5d80-6eca-11ea-9b83-06407f11be2e.png">

![image](https://user-images.githubusercontent.com/77539/77704586-2c7a1100-6f9c-11ea-830d-a54d6d70e985.png)

### Technical implementation

In terms of technical implementation, the main goal is to keep the universality of the block as much as possible. It means that colors are handled by the same code for both sides of the app: editor-canvas and front-end.
For this purpose, I've moved the attributes into the `attributes` property of the `<PodcastPlayer />` component:

Inspecting properties in the **editor-canvas**:
![image](https://user-images.githubusercontent.com/77539/77605317-5b887800-6ef3-11ea-9d0f-daf6a6d9d10b.png)

and in the **front-end**:
![image](https://user-images.githubusercontent.com/77539/77605351-6fcc7500-6ef3-11ea-9428-9de03d283a59.png)

Fixes https://github.com/Automattic/jetpack/issues/14940

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
*

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here.

#### Testing instructions:

* Apply the patch.
* Set colors through of the inspector controls in the sidebar. Confirm that colors are stored and shown in the front-end.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* Add `primary`, `secondary` and `background` colors in the inspector controls.
* Move attributes into its own field.
* Apply colors on both sides via client-side scrips.
* Implement contrast checker between secondary and background color.

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: declare colors attributes

retrofox committed:
[not verified] [not verified] [not verified] extension: add withFallbackStyles HoC

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: compose colors into the edit cmp

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: add colors to inspector controls

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: set colors via CSS classes

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: apply colors via inline style

retrofox committed:
[not verified] [not verified] [not verified] [not verified] podcast-plater: handling colors style

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: rename text color to primary color

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: add secondary color

retrofox committed:
[not verified] podcats-player: move attributes into its field

retrofox committed:
[not verified] [not verified] podcast-player: refact applying colors in edit
This commit delegates the behavior to apply colors to the podcast player in order to keep the universality as much as possible.

retrofox committed:
[not verified] podcast-player: apply colors in PodcastPlayer cmp

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: propagate primary colors

retrofox committed:
podcast-player: apply CSS classes to actve podcast

retrofox committed:
[not verified] podcast-player: apply secondary color to <li>s

retrofox committed:
[not verified] [not verified] [not verified] podcast-player: apply inline style to main el

retrofox committed:
[not verified] [not verified] podcast-player: apply inline styles to podcasts

marekhrabe commented:
Color options work well for me when testing this PR. You have also made a good point about `attributes` prop - it should work well and I agree it will be simpler to add a new attribute with it.

As for the technical approach, I have tried to [warn you against using editor dependencies](https://github.com/Automattic/jetpack/pull/14991#issuecomment-603432103) in `PodcastPlayer` and you still have included here. This means a few things I have tried hard to avoid:
- we don't have any color handling on the backend, meaning we can't code a markup that would look the same as the React component. We need this in order to be able to
  - show player completely without JavaScript, respecing block settings from attributes
  - show player in a browser with JavaScript but before React and our view.js loads (because we load them asynchronously in the footer of the page)
- the main blocker I have: we are no longer including just React and its `@wordpress/element` wrapper but also the whole block editor itself and its dependencies :( Those are obviously available at no cost in the editor but on the frontend that's extra megabytes of JavaScript.

These are the new scripts that are being included from your PR on the frontend (all from `/wp-includes/js/dist/`:

dom-ready.js, a11y.js, blob.js, autop.js, block-serialization-default-parser.js, is-shallow-equal.js, compose.js, hooks.js, deprecated.js, priority-queue.js, redux-routine.js, data.js, dom.js, html-entities.js, shortcode.js, blocks.js, vendor/moment.js, keycodes.js, rich-text.js, components.js, token-list.js, url.js, viewport.js, wordcount.js, block-editor.js

Loading a post on the frontend with Podcast Player included looks like this on my site:
`master`
<img width="312" alt="Screenshot 2020-03-26 at 10 09 57" src="https://user-images.githubusercontent.com/156676/77629673-31ae7000-6f4a-11ea-8c1e-071621680be0.png">

`update/podcast-player-colors`
<img width="322" alt="Screenshot 2020-03-26 at 10 10 09" src="https://user-images.githubusercontent.com/156676/77629692-36732400-6f4a-11ea-8a6e-6f9f359b0bf6.png">

getdave commented:
@retrofox I was reading what @marekhrabe wrote and I hoped that perhaps we might find that Webpack tree shaking strips out the pieces of `@wordpress/block-editor` which we don't use. But it would probably only do this in "production" build mode. We'd need to ask Jetpack Crew about this.

For now, I've simply copied the implementation of the only block editor function you are using `getColorClassName` into a `util` within our Block.

I'll leave the rest to you. I just wanted to show that it might be possible to work around this. Still needs the same implementation on the PHP side.

marekhrabe commented:
> Webpack tree shaking strips out the pieces of @wordpress/block-editor which we don't use. But it would probably only do this in "production" build mode. We'd need to ask Jetpack Crew about this.

We've talked a bit more about it on Slack. This is expected as there is no tree shaking and we are not using webpack to require anything. Webpack removes those `@wordpress/*` imports, replaces them with reading `window.wp.packageName` directly and adds a dependency to your `.asset.php`

If your dependency is `block-editor`, then all its dependencies will have to the included on a page too

retrofox commented:
> If your dependency is `block-editor`, then all its dependencies will have to the included on a page too

Removing this dependency.

retrofox commented:
> For now, I've simply copied the implementation of the only block editor function you are using `getColorClassName` into a `util` within our Block.

Thanks., and yes, that makes totally sense, since it's the only dependency that I've added in this PR. Pretty easy to remove. I Need to rebase because the branch got conflicts again. in order to apply the patch.

retrofox commented:
> We've talked a bit more about it on Slack.

Could you point me where? Please next time feel free to ping me since they seem to be important decisions about what I'm working on now. 🙇

retrofox commented:
> * we don't have any color handling on the backend, meaning we can't code a markup that would look the same as the React component. We need this in order to be able to
>   
>   * show player completely without JavaScript, respecing block settings from attributes
>   * show player in a browser with JavaScript but before React and our view.js loads (because we load them asynchronously in the footer of the page)

Fair enough, very valid point. Going to keep working on the server-side.
The reason why I opted to apply the colors in the client-side if some of them depend on the user interaction, and it isn't doable on the server-side.

> * the main blocker I have: we are no longer including just React and its `@wordpress/element` wrapper but also the whole block editor itself and its dependencies :( Those are obviously available at no cost in the editor but on the frontend that's extra megabytes of JavaScript.

I'm removing the dependency. That's another very good point.

> 
> These are the new scripts that are being included from your PR on the frontend (all from `/wp-includes/js/dist/`:
> 
> dom-ready.js, a11y.js, blob.js, autop.js, block-serialization-default-parser.js, is-shallow-equal.js, compose.js, hooks.js, deprecated.js, priority-queue.js, redux-routine.js, data.js, dom.js, html-entities.js, shortcode.js, blocks.js, vendor/moment.js, keycodes.js, rich-text.js, components.js, token-list.js, url.js, viewport.js, wordcount.js, block-editor.js
> 
> Loading a post on the frontend with Podcast Player included looks like this on my site:
> `master`
> <img alt="Screenshot 2020-03-26 at 10 09 57" width="312" src="https://user-images.githubusercontent.com/156676/77629673-31ae7000-6f4a-11ea-8c1e-071621680be0.png">
> 
> `update/podcast-player-colors`
> <img alt="Screenshot 2020-03-26 at 10 10 09" width="322" src="https://user-images.githubusercontent.com/156676/77629692-36732400-6f4a-11ea-8a6e-6f9f359b0bf6.png">

The performance analysis is very good and useful. Thanks for taking part of your time on this.
Maybe it also evidences that we are taking an approach that could fall off on this problem easily. The issues described above was caused for this single line in the code source:

```es6
import { getColorClassName } from '@wordpress/block-editor';
```
In this case, it's going to be fixed because what the `getColorClassName()` function does is not a big deal, pretty easy to replace and even @getdave made a commit suggestion for this (btw it was missed in the rebasing process)

Now wondering how we are going to prevent these potential issues in the future? Adding documentation at the top to each file that is involved in the React rendering for the front-end side is something that we could do, but there isn't guaranteed that in the future it wasn't going to occur.

retrofox committed:
[not verified] [not verified] podcast-player: add util.js file
It includes the getColorClassName() function in order to avoid including a bunch of libraries when it was previously included by @wordpress/block-editor lib.

retrofox committed:
[not verified] podcast-player: fix handling list styling issue

marekhrabe commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -203,20 +219,55 @@ const PodcastPlayerEdit = ( {
+					<ContrastChecker
+						{ ...{
+							// Text is considered large if font size is greater or equal to 18pt or 24px,
+							// currently that's not the case for button.
```
Is the button reference meant to be here or is it just copied from somewhere?

retrofox replied:
copied from elsewhere, product if a wrong rebase I guess 🤦‍♂ . Updating

retrofox replied:
but the comment there makes sense jus in case.

marekhrabe commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -203,20 +219,55 @@ const PodcastPlayerEdit = ( {
+						},
+					] }
+				>
+					<ContrastChecker
```
I wasn't able to trigger the contrast checker on my testing site. What's the best way to test this? I have tried setting all three colors to the same custom color (`#fff`) and it still didn't show up for me.

retrofox replied:
let me take a look. Also, I think we'll have to iterate on it.

marekhrabe commented on extensions/blocks/podcast-player/components/playlist.js:
```diff
@@ -1,9 +1,10 @@
- * External dependencies
+ * WordPress dependencies
  */
 import classnames from 'classnames';
```
classnames should be external

retrofox replied:
yes 👍 Updating

marekhrabe commented on extensions/blocks/podcast-player/style.scss:
```diff
@@ -58,9 +57,17 @@ $block-border-color: $dark-gray-100;
+	&.is-active.has-primary .jetpack-podcast-player__episode-link,
+	&.has-secondary .jetpack-podcast-player__episode-link {
+		color: inherit;
+	}
```
Could you please add a short comment that explains it? I think it will make it easier for the next person to grasp what's happening. It took me a bit to mentally parse it 😅

retrofox replied:
👍

retrofox replied:
Also, we'll have to think a nice way to define default colors, since they can be handled by CSS or through the clock properties., or both of them.

marekhrabe commented:
This is working really well for me. Any combination of named or custom colors for those fields works in the editor and the frontend too. Color selections are properly saved, too. 

I have left a few comments but those are very minor details, except for the contrast checker. Any idea what's happening there?

retrofox committed:
[not verified] [not verified] podcast-player: fix contrast checker
It checks the contrast between the secondy and the background color at the moment. We can iterate later in a next PR.

retrofox committed:
[not verified] podcast player: improve doc

retrofox committed:
[not verified] [not verified] podcast-player: sort dependencies

getdave commented on extensions/blocks/podcast-player/components/playlist.js:
```diff
@@ -2,13 +2,18 @@
  * Internal dependencies
  */
 import { STATE_ERROR } from '../constants';
+import { getColorClassName } from '../util';
```
```suggestion
import { getColorClassName } from '../utils';
```

Could we rename this to plural. It's likely we may end up with more than one. It's odd. I know it was my fault - sorry!

retrofox replied:
<3

getdave commented on extensions/blocks/podcast-player/components/playlist.js:
```diff
@@ -25,69 +30,113 @@ const TrackError = memo( ( { link } ) => (
+						primaryColor={ primaryColor }
+						customPrimaryColor={ customPrimaryColor }
+						secondaryColor={ secondaryColor }
+						customSecondaryColor={ customSecondaryColor }
```
Just a thought. This number of props on this component is getting quite large.

I wonder whether a single `colors` prop would be better.

Something like

```js
{
    'primary': {
        'default': '',
        'custom': '',
    }
    'secondary': {
        'default': '',
        'custom': '',
    }
}
```

retrofox replied:
sounds good to me. Let me update the code.

getdave commented on extensions/blocks/podcast-player/podcast-player.php:
```diff
@@ -107,8 +107,8 @@ function render_player( $player_data, $attributes ) {
 	$player_props = array_merge(
-		// Make all attributes available.
-		$attributes,
+		// Add all attributes.
```
Should we explain _why_ we need to nest attributes? What was the motivation?

retrofox replied:
I tried to explain it in the PR description. Basically I'd like to put the block attributes into its own field in order to avoid the mix them with the block properties. From my point fo view, and after working on this, it feels cleaner and simpler to follow and understand,

getdave reviewed (commented):
Noticed some minor code points. Haven't tested yet, but looks like @marekhrabe beat me to that.

jeherve commented on extensions/blocks/podcast-player/attributes.js:
```diff
@@ -20,4 +24,22 @@ export default {
 		default: true,
 	},
+	primaryColor: {
+		type: 'string',
```
Do you think we could add our `colorValidator` utility as a `validator` for those colors, if we always expect hex values?

retrofox replied:
`primaryColor` is a string. We could add hex values for the custom ones, though. Let me take a look at how to do it. Good advice.

jeherve commented on extensions/blocks/podcast-player/edit.js:
```diff
@@ -203,20 +219,51 @@ const PodcastPlayerEdit = ( {
-					itemsToShow={ itemsToShow }
-					showEpisodeDescription={ showEpisodeDescription }
-					showCoverArt={ showCoverArt }
+					coverArt={ feedData.coverArt }
```
Should it be `showCoverArt` here, to match uses elsewhere?

retrofox replied:
If I'm not wrong `coverArt` is the source of the cover art, while `showCoverArt` is the block attribute which defines if the cover art should be shown or not.

marekhrabe replied:
We use plain `cover` in API and all components for a while now to hold the image cover URL. You can see it a few lines above:

`cover={ feedData.cover }`

this `coverArt` line should be removed.

jeherve reviewed (changes requested):
This works quite nicely. Left a few remarks below.

For some reason, the default settings already trigger the contrast warning with Twenty Twenty:

![image](https://user-images.githubusercontent.com/426388/77684250-ff2c6380-6f99-11ea-8abd-819458e2c4d1.png)

This doesn't seem ideal. Is there something we can do about it?

retrofox committed:
[not verified] [not verified] podcast-player: rename utils.js

since maybe there more than utils, let's rename to utils.js

Co-Authored-By: Dave Smith <getdavemail@gmail.com>

retrofox committed:
[not verified] [not verified] podcast-player: finish renaming utils

retrofox committed:
[not verified] [not verified] podcast-player: refactoring colors props

retrofox committed:
[not verified] podcast-player: handling secondary color with CSS

getdave commented on extensions/blocks/podcast-player/utils.js:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Returns a class based on the context a color is being used and its slug.
```
Might be worth mentioning why we are not using the one from Gutenberg Core here just in case.

retrofox replied:
> > What do you think about kicking it to another next one?
> 
> Sure thing.

Done some fixes [here](https://github.com/Automattic/jetpack/pull/15110/commits/a26dc10a3b1cc6788638d438da76e6e79a333061).

retrofox commented:
> This works quite nicely. Left a few remarks below.
> 
> For some reason, the default settings already trigger the contrast warning with Twenty Twenty:
> 
> ![image](https://user-images.githubusercontent.com/426388/77684250-ff2c6380-6f99-11ea-8abd-819458e2c4d1.png)
> 
> This doesn't seem ideal. Is there something we can do about it?

Definitely we'll have to keep working on the contrast checker, but I suggest do it in a new PR.
The main reason we need to re-think this section considering that now there are three kinds of colors: `primary`, `secondary` and `background`. 
What do you think about kicking it to another next one?

getdave reviewed (dismissed):
<img width="2555" alt="Screenshot 2020-03-27 at 11 02 15" src="https://user-images.githubusercontent.com/444434/77749747-73a6e700-701a-11ea-8207-684ce85377d3.png">

I tested this and the output is as expected across editor and frontend view. Looking really good.

Thanks for working on all the feedback. I feel this in a much better place.

jeherve commented:
> What do you think about kicking it to another next one?

Sure thing.

retrofox commented:
> > What do you think about kicking it to another next one?
> 
> Sure thing.

👀 now, I guess there is something wrong.

retrofox committed:
[not verified] podcast-player: fix geet fallbak colors

retrofox committed:
[not verified] podcast-player: validate custom colors

retrofox committed:
[not verified] podcast-player: improve getColorClassName jsdoc

retrofox committed:
[not verified] podcast-player: use colorValidator() helper

retrofox commented:
@jeherve I've updated the code according to folks suggestions, such as using colorValidator(), fix some issues with the contrast-checker beyond we'll still keep working on this in the near future, etc.

Thanks for the review.

marekhrabe commented on extensions/blocks/podcast-player/icons.js:
```diff
@@ -0,0 +1,43 @@
+/**
```
This was removed in master. Please remove it here so it doesn't get added back.

marekhrabe commented:
I have tested this and everything works well ✅Happy to approve once an extraneous file and one line of code get deleted, as noted in review comments.

retrofox committed:
podcast-player: remove coverArt line

retrofox committed:
podcast-player: remove icons.js file

retrofox commented:
> I have tested this and everything works well ✅Happy to approve once an extraneous file and one line of code get deleted, as noted in review comments.

Thanks, already updated.

jeherve reviewed (approved):
This looks good, it should be good to merge.

Quick note, since the PR introduces a new file, you'll need to update the matching WordPress.com diff manually.

retrofox commented:
r205000-wpcom
